lib/gtfs.ex

```diff
@@ -0,0 +1,165 @@
+defmodule Gtfs do
+  use GenServer
+  require Logger
+
+  @type t :: %__MODULE__{
+          routes: [Gtfs.Route.t()],
+          stop_times: [Gtfs.StopTime.t()],
+          stops: [Gtfs.Stop.t()],
+          trips: [Gtfs.Trip.t()]
+        }
+
+  @enforce_keys [
+    :routes,
+    :stop_times,
+    :stops,
+    :trips
+  ]
+
+  defstruct [
+    :routes,
+    :stop_times,
+    :stops,
+    :trips
+  ]
+
+  @type state :: t
+
+  @type query_opts :: %{
+          optional(:server) => GenServer.server()
+        }
+
+  @type files :: %{optional(String.t()) => binary()}
+
+  @type files_source :: {:url, String.t()} | {:mocked_files, mocked_files()}
+
+  @typedoc """
+  For mocking tests
+  E.g.
+  %{
+    "stops.txt" => [
+      "stop_id,stop_name",
+      "place-sstat,South Station",
+    ]
+  ]}
+  """
+  @type mocked_files :: %{optional(String.t()) => [binary()]}
+
+  # Client functions
+
+  @spec start_link(String.t()) :: GenServer.on_start()
+  def start_link(url) do
+    GenServer.start_link(__MODULE__, {:url, url}, name: __MODULE__)
+  end
+
+  @spec start_mocked(mocked_files) :: GenServer.on_start()
+  def start_mocked(mocked_files) do
+    GenServer.start_link(__MODULE__, {:mocked_files, mocked_files})
+  end
+
+  @doc """
+  Returns all of GTFS, with each file as an item in a map
+  """
+  @spec gtfs(query_opts()) :: t()
+  def gtfs(opts) do
+    server = opts[:server] || __MODULE__
+    GenServer.call(server, :state)
+  end
+
+  # GenServer callbacks
+
+  @impl true
+  def init(files_source) do
+    case fetch_files(files_source) do
```

Comment:
This works, but `with` is another option for this as well (and same in `fetch_files`).

Reply:
`with` will return the failing value, right? In this case, we want control over that value, so we can return `{:stop, ...}` instead of whatever `HTTPoison` returns.

Reply:
You can use `else` with `with`, so you'd have something along the lines of:

    with {:files, files} <- fetch_files(files_source) do
      <do work>
    else
      {:err, err} ->
        {:stop, err}
    end

(Optional, but returning data as a tuple (i.e. `{:files, files}`) can make it better for deciphering positive results from errors in some cases.)

Reply:
You can return `:not_loaded` or an empty list if the data hasn't loaded yet. It's an OTP anti-pattern to do long processing in `init/1`, as the supervisor can't do anything while it waits for the process to start. See https://medium.com/@derek.kraan2/avoid-these-otp-supervision-performance-pitfalls-c3fe64e70f9e for more.

Reply:
This article alleviates my concern that loading asynchronously would give other processes incomplete information: https://elixirschool.com/blog/til-genserver-handle-continue/

> This approach ensures that our GenServer won’t handle any other messages until `handle_continue/2` is finished. No more race conditions!